nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/FetchAzureBlobStorage.java

```diff
@@ -103,10 +107,15 @@ public void onTrigger(ProcessContext context, ProcessSession session) throws Pro
             session.transfer(flowFile, REL_SUCCESS);
             final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
             session.getProvenanceReporter().fetch(flowFile, blob.getSnapshotQualifiedUri().toString(), transferMillis);
-
-        } catch (IllegalArgumentException | URISyntaxException | StorageException e1) {
-            flowFile = session.penalize(flowFile);
-            session.transfer(flowFile, REL_FAILURE);
+        } catch (IllegalArgumentException | URISyntaxException | StorageException | ProcessException e) {
+            if (e instanceof ProcessException && storedException.get() == null) {
+                throw (ProcessException) e;
+            } else {
+                Exception failureException = Optional.ofNullable(storedException.get()).map(x -> x).orElse(e);
```

Comment:
@jtstorck I don't think you need the .map() here.  It looks like [orElse()](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#orElse-T-) will already give you the value if the optional is absent.